### PR TITLE
Add support for custom endpoints in the Azure repository

### LINKED ISSUE #94537: Add support for custom Azure endpoints in azure repository
Currently we only support adding an endpoint suffix for azure repositories. This is not enough when a custom Azure DNS  zone is used since the URI that's created using an endpoint suffix is incorrect.
```java
primaryUri = new URI(protocol + "://" + accountName + "." + service + "." + endpointSuffix);
``` 

We should add the ability to provide a custom endpoint to cover this situations.

### PR DESCRIPTION
**Pull request by fcofdez**
Closes #94537

### CODE REVIEW

**Review by idegtiarenko — APPROVED** [review]

File: docs/reference/snapshot-restore/repository-azure.asciidoc [hunk 1]
```diff
@@ -140,6 +140,12 @@ stored in the keystore are marked as "secure"; the other settings belong in the
   set by the Azure client (known as 5 minutes). This setting can be defined
   globally, per account, or both.
 
+`endpoint`::
+  The Azure endpoint to connect to. It must include the protocol used to connect to Azure.
+
+`secondary_endpoint`::
```
Comment (arteam) [inline docs/reference/snapshot-restore/repository-azure.asciidoc:146]: I've been trying to find the docs about the use case for the secondary endpoint, but couldn't find any specifics in Microsoft docs. Is it needed for redundancy?

File: modules/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureStorageServiceTests.java [hunk 2]
```diff
@@ -450,6 +450,71 @@ public void testInvalidSettingsRetryConfigurationForLocationModeWithSecondaryFal
...
+
+            expectThrows(IllegalArgumentException.class, () -> azureStorageService.client("azure1", LocationMode.PRIMARY_THEN_SECONDARY));
+            expectThrows(IllegalArgumentException.class, () -> azureStorageService.client("azure1", LocationMode.SECONDARY_ONLY));
+            expectThrows(IllegalArgumentException.class, () -> azureStorageService.client("azure1", LocationMode.SECONDARY_THEN_PRIMARY));
+
+            AzureBlobServiceClient client1 = azureStorageService.client("azure1", LocationMode.PRIMARY_ONLY);
+            assertThat(client1.getSyncClient().getAccountUrl(), equalTo("https://account1.zone.azure.net"));
+
+            {
+                AzureBlobServiceClient client = azureStorageService.client(
```
Comment (arteam) [inline modules/repository-azure/src/test/java/org/elasticsearch/repositories/azure/AzureStorageServiceTests.java:472]: Nit: I think we can avoid declaring the `client` variable and declaring a code block to void scoping issues.
```java
assertThat(
    azureStorageService.client("azure2", randomBoolean() ? LocationMode.PRIMARY_ONLY : LocationMode.PRIMARY_THEN_SECONDARY)
        .getSyncClient()
        .getAccountUrl(),
    equalTo("https://account2.zone.azure.net")
);
```

**Review by arteam — APPROVED** [review]
LGTM! I've left a few small comments.

File: docs/reference/snapshot-restore/repository-azure.asciidoc (hunk 1 above)
Comment (fcofdez) [reply docs/reference/snapshot-restore/repository-azure.asciidoc:146]: It's used for retries, see https://learn.microsoft.com/en-us/azure/storage/blobs/storage-retry-policy#use-geo-redundancy-to-improve-app-resiliency

File: docs/reference/snapshot-restore/repository-azure.asciidoc [hunk 3]
```diff
@@ -140,6 +140,12 @@ stored in the keystore are marked as "secure"; the other settings belong in the
   set by the Azure client (known as 5 minutes). This setting can be defined
   globally, per account, or both.
 
+`endpoint`::
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply docs/reference/snapshot-restore/repository-azure.asciidoc:143]: Do we have something similar in other repositories?

**Pull request merged**